Controlled alignfull padding differently to correct paragraph padding in certain situation.

pbking opened the pull request:
#### Changes proposed in this Pull Request:
This changes the padding rules around groups (particularly full width groups) so that paragraph padding rules behave as expected.

The result is that ALL group blocks that are configured to `alignfull` introduce a standard-width padding.  This changes the behavior of some blocks when inside of a group block (such as an image) which used to expand fully (only if the original image was wide enough).

Before:
<img src="https://user-images.githubusercontent.com/146530/123834637-a0107980-d8d5-11eb-87f1-2f254664e713.png" width="500">
<img src="https://user-images.githubusercontent.com/146530/123834556-896a2280-d8d5-11eb-831a-786b78320a18.png" width="300">


After:
<img src="https://user-images.githubusercontent.com/146530/123834382-5e7fce80-d8d5-11eb-9579-be22790dfed5.png" width="500">
<img src="https://user-images.githubusercontent.com/146530/123834456-70617180-d8d5-11eb-835a-4aa8f89a508f.png" width="300">


#### Related issue(s):
Fixes #4099

pbking committed:
Controlled alignfull padding differently to correct paragraph padding in certain situations.

jffng reviewed (commented):
I tried out the markup @kjellr supplied and ran into a case that this change doesn't quite cover:

```
<!-- wp:query {"queryId":15,"query":{"perPage":1,"pages":0,"offset":0,"postType":"post","categoryIds":[],"tagIds":[],"order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false},"align":"full"} -->
<div class="wp-block-query alignfull"><!-- wp:post-template -->
<!-- wp:group {"layout":{"inherit":true}} -->
<div class="wp-block-group"><!-- wp:post-title {"isLink":true} /-->

<!-- wp:post-featured-image {"isLink":true} /-->

<!-- wp:post-excerpt /-->

<!-- wp:post-date /--></div>
<!-- /wp:group -->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```

In the above example, you would need to add `alignfull` to the group to get the correct padding on mobile. This feels a bit counterintuitive, but I'm not sure of a better approach because we don't want to add the padding to _all_ group blocks — this would interfere with the layout control. 

Considering the plethora of ideas we tried in our 🍐  session, I think this is the simplest approach that integrates well with the existing layout controls. 

Curious if others have ideas though; it's a tricky problem.

pbking commented:
I think that in the instance you raised that the `<!-- wp:group -->` block isn't actually needed at all to achieve the design target, and eliminating it makes the situation simpler. Remove `alignFull` from the `wp:query` block and the content get's the regular margins...

Maybe this borks up something else though like featured images that I haven't considered in my limited testing...

```
<!-- wp:query {"queryId":15,"query":{"perPage":5,"pages":0,"offset":0,"postType":"post","categoryIds":[],"tagIds":[],"order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:post-title {"isLink":true} /-->

<!-- wp:post-featured-image {"isLink":true} /-->

<!-- wp:post-excerpt /-->

<!-- wp:post-date /-->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```

kjellr commented:
@pbking I think this PR is missing one additional rule — when layout _is_ active, `alignfull` children should extend all the way to the edges. This is a quick fix, you can just modify [the rules above your edits](https://github.com/Automattic/themes/blob/f0422058c29978ab507a51adbc060bd3e5f0ab30/blockbase/sass/base/_alignment.scss#L8-L13) to add a couple new selectors: 

```
.block-editor-block-list__layout.is-root-container>.wp-block[data-align=full],
.wp-block-post-content > .alignfull,
[data-align="full"] .wp-block-group > [data-align="full"],
.wp-block-group.alignfull > .alignfull {
	margin-left: calc(-1 * var(--wp--custom--post-content--padding--left)) !important;
	margin-right: calc(-1 * var(--wp--custom--post-content--padding--right)) !important;
	width: calc( 100% + var(--wp--custom--post-content--padding--left) + var(--wp--custom--post-content--padding--right) ) !important;
}
```

Before|After
---|---
<img width="999" alt="Screen Shot 2021-07-01 at 10 31 19 AM" src="https://user-images.githubusercontent.com/1202812/124141540-7e3f0000-da57-11eb-8420-54161c698514.png">|<img width="998" alt="Screen Shot 2021-07-01 at 10 30 12 AM" src="https://user-images.githubusercontent.com/1202812/124141547-7ed79680-da57-11eb-8063-3e349cd59e18.png">

kjellr commented:
> Maybe this borks up something else though like featured images that I haven't considered in my limited testing...

Yeah, wide and full alignments within the block won't work with this setup. This featured image is supposed to be wide-width for example: 

<img width="1107" alt="Screen Shot 2021-07-01 at 10 32 59 AM" src="https://user-images.githubusercontent.com/1202812/124142136-f7d6ee00-da57-11eb-9ead-01299881dc8e.png">

jffng commented:
I gave this another spin and verified a couple things:

- @kjellr's suggestion plays nicely with the changes already in this PR. 
- It still requires that the group block inside the query has "alignfull" for the padding to be applied correctly on mobile. 

Something else worth mentioning — we can't solve this problem with a media query because we don't know the viewport width to add the rule, given how the layout system works.

pbking committed:
Refactored 'container' alignment rules

Eliminated 'page-content' class as a means to apply default 'edge'
padding to content.

Introduced 'container-' selection as the means to apply default 'edge'
padding to content.

Removed 'edge' padding from post-content block (it is all handled by
'container-' elements now).

Refactored the "bust out of edge padding" for alignfull elements and
nested 'container-'s.

pbking committed:
Demonstrate the ability to declare full width stylings in a query loop

pbking commented:
Shooooooo!

@jffng and I gave this ANOTHER another spin around the block and then on the freeway and then around the block again and... did some more refactoring.

The result was the following:

Eliminated ['page-content'](https://github.com/Automattic/themes/pull/4123/commits/3fc199666a81d67f1d818e7fd96517239fe75478#diff-48d2e95a15120f192001e9c23b01a6e0d8609a31615b24ff2f30e2723a8d2389L17) [class](https://github.com/Automattic/themes/pull/4123/commits/3fc199666a81d67f1d818e7fd96517239fe75478#diff-48d2e95a15120f192001e9c23b01a6e0d8609a31615b24ff2f30e2723a8d2389L17) as a means to apply default 'edge'
padding to content.

Introduced ['container-' selection](https://github.com/Automattic/themes/pull/4123/commits/3fc199666a81d67f1d818e7fd96517239fe75478#diff-48d2e95a15120f192001e9c23b01a6e0d8609a31615b24ff2f30e2723a8d2389R9) as the means to apply default 'edge'
padding to content.

Removed 'edge' padding from[ post-content block ](https://github.com/Automattic/themes/pull/4123/commits/3fc199666a81d67f1d818e7fd96517239fe75478#diff-8d19c6d8b05bdcca09ecf489813da09999af47dd7cfafc71b7e263f900b97410L378)(it is all handled by
'container-' elements now).

Refactored the ["bust out of edge padding"](https://github.com/Automattic/themes/pull/4123/commits/3fc199666a81d67f1d818e7fd96517239fe75478#diff-48d2e95a15120f192001e9c23b01a6e0d8609a31615b24ff2f30e2723a8d2389R15-R22) for alignfull elements and
nested 'container-'s.

Lastly we [demonstrated](https://github.com/Automattic/themes/pull/4123/commits/25ca5699b5ee459e0ac11d782c09406ba194b5a6) one of the more complex scenarios this allow for (permitting an element to 'alignfull' inside of a query block loop).  This change is not aligned with the expected design, but the attempt and resolution of that behavior seemed important to demonstrate.  That should be reverted once evaluation is complete.

There's a TON of behavior that we did NOT test while working on this that should be evaluated for correctness.  There aren't many lines changed but all-in-all it's a pretty heft alignments change that should be given heavy scrutiny.

pbking committed:
Refactored Mayland-blocks to work with Blockbase's modified alignment rules

scruffian reviewed (approved):
This seems to work in my tests. I'd still love for us to use grid in core...

pbking committed:
Removed demo code
